File: frontend/src/lib/esl.ts
import { apiFetch } from "@/lib/api";
import type { Shipment } from "@/lib/shipments";

export type TransportType = "SHIP" | "TRUCK";
export type TransportSlotStatus = "AVAILABLE" | "BOOKED" | "MAINTENANCE";

export type TransportSlot = {
  id: string;
  transport_type: TransportType;
  name: string;
  reference_code: string;
  origin: string;
  destination: string;
  capacity_kg: string;
  remaining_capacity_kg: string;
  capacity_cbm?: string;
  remaining_capacity_cbm?: string;
  available_from: string;
  status: TransportSlotStatus;
  created_at: string;
  updated_at: string;
};

export type ShipmentAllocation = {
  id: string;
  shipment_id: string;
  transport_slot_id: string;
  leg_type: "SEA" | "INLAND";
  esl_agent_id: string;
  expected_departure_at: string;
  notes?: string;
  confirmed_at: string;
  created_at: string;
};

export type ShipmentAllocationDetail = {
  allocations: ShipmentAllocation[];
  shipment: Shipment;
  slots: TransportSlot[];
};

export async function listVerifiedShipmentsForAllocation(
  token: string,
): Promise<Shipment[]> {
  const res = await apiFetch<{ items?: Shipment[] | null }>(
    "/api/v1/esl/shipments/verified",
    { token },
  );
  return Array.isArray(res.items) ? res.items : [];
}

export async function listAvailableTransportSlots(
  token: string,
): Promise<TransportSlot[]> {
  const res = await apiFetch<{ items?: TransportSlot[] | null }>(
    "/api/v1/esl/transport-slots",
    { token },
  );
  return Array.isArray(res.items) ? res.items : [];
}

export function allocateShipment(
  token: string,
  shipmentID: string,
  payload: {
    ship_slot_id: string;
    truck_slot_id: string;
    expected_departure_date: string;
    notes?: string;
  },
): Promise<ShipmentAllocationDetail> {
  return apiFetch<ShipmentAllocationDetail>(
    `/api/v1/esl/shipments/${shipmentID}/allocate`,
    {
      method: "POST",
      token,
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    },
  );
}

File: backend/internal/domain/esl.go
package domain

import "time"

type TransportType string

const (
	TransportTypeShip  TransportType = "SHIP"
	TransportTypeTruck TransportType = "TRUCK"
)

type TransportSlotStatus string

const (
	TransportSlotAvailable   TransportSlotStatus = "AVAILABLE"
	TransportSlotBooked      TransportSlotStatus = "BOOKED"
	TransportSlotMaintenance TransportSlotStatus = "MAINTENANCE"
)

type TransportSlot struct {
	ID                   string              `json:"id"`
	TransportType        TransportType       `json:"transport_type"`
	Name                 string              `json:"name"`
	ReferenceCode        string              `json:"reference_code"`
	Origin               string              `json:"origin"`
	Destination          string              `json:"destination"`
	CapacityKG           string              `json:"capacity_kg"`
	RemainingCapacityKG  string              `json:"remaining_capacity_kg"`
	CapacityCBM          string              `json:"capacity_cbm,omitempty"`
	RemainingCapacityCBM string              `json:"remaining_capacity_cbm,omitempty"`
	AvailableFrom        time.Time           `json:"available_from"`
	Status               TransportSlotStatus `json:"status"`
	CreatedAt            time.Time           `json:"created_at"`
	UpdatedAt            time.Time           `json:"updated_at"`
}

type ShipmentAllocation struct {
	ID                  string    `json:"id"`
	ShipmentID          string    `json:"shipment_id"`
	TransportSlotID     string    `json:"transport_slot_id"`
	LegType             string    `json:"leg_type"`
	ESLAgentID          string    `json:"esl_agent_id"`
	ExpectedDepartureAt time.Time `json:"expected_departure_at"`
	Notes               string    `json:"notes,omitempty"`
	ConfirmedAt         time.Time `json:"confirmed_at"`
	CreatedAt           time.Time `json:"created_at"`
}

type ShipmentAllocationDetail struct {
	Allocations []ShipmentAllocation `json:"allocations"`
	Shipment    Shipment             `json:"shipment"`
	Slots       []TransportSlot      `json:"slots"`
}
